buildkite/resource_team_member.go

```diff
@@ -3,174 +3,208 @@ package buildkite
-				Type:     schema.TypeString,
-				ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
-					v := val.(string)
-					switch v {
-					case "MEMBER":
-					case "MAINTAINER":
-						return
-					default:
-						errs = append(errs, fmt.Errorf("%q must be either MEMBER or MAINTAINER, got: %s", key, v))
-						return
-					}
-					return
+func newTeamMemberResource() resource.Resource {
+	return &teamMemberResource{}
+}
+
+func (teamMemberResource) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_team_member"
+}
+
+func (tm *teamMemberResource) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	tm.client = req.ProviderData.(*Client)
+}
+
+func (teamMemberResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = resource_schema.Schema{
+		MarkdownDescription: "A team member resourc allows for the management of team team membership for existing organization users.",
+		Attributes: map[string]resource_schema.Attribute{
+			"id": resource_schema.StringAttribute{
+				Computed: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
 				},
 			},
-			"team_id": &schema.Schema{
-				Required: true,
-				Type:     schema.TypeString,
+			"uuid": resource_schema.StringAttribute{
+				Computed: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
 			},
-			"user_id": &schema.Schema{
-				Required: true,
-				Type:     schema.TypeString,
+			"team_id": resource_schema.StringAttribute{
+				Required:            true,
+				MarkdownDescription: "The GraphQL ID of the team to add/remove the user to/from.",
 			},
-			"uuid": &schema.Schema{
-				Computed: true,
-				Type:     schema.TypeString,
+			"user_id": resource_schema.StringAttribute{
+				Required:            true,
+				MarkdownDescription: "The GraphQL ID of the user to add/remove.",
+			},
+			"role": resource_schema.StringAttribute{
+				Optional:            true,
+				MarkdownDescription: "The role for the user. Either MEMBER or MAINTAINER.",
+				Validators: []validator.String{
+					stringvalidator.OneOf("MEMBER", "MAINTAINER"),
+				},
 			},
 		},
 	}
 }
 
-// CreateTeamMember adds a user to a Buildkite team
-func CreateTeamMember(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
-	originalRole := d.Get("role").(string)
-	client := m.(*Client)
-
-	var mutation struct {
-		TeamMemberCreate struct {
-			TeamMemberEdge struct {
-				Node TeamMemberNode
-			}
-		} `graphql:"teamMemberCreate(input: {teamID: $teamId, userID: $userId})"`
-	}
+func (tm *teamMemberResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan, state teamMemberResourceModel
```

Comment:
If it's just a Create, you don't need 2 iterations of the Model. Maybe just use `plan` for clarity as on Create there is no `state`.

Reply:
Dropping to one model in the latest commit